File: library/src/main/java/com/snail/easyble/core/Device.java
package com.snail.easyble.core;

import android.bluetooth.BluetoothDevice;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

/**
 * 描述: 蓝牙设备
 * 时间: 2018/4/11 15:00
 * 作者: zengfansheng
 */
public class Device implements Comparable<Device>, Cloneable, Parcelable {
    /**
     * 原始设备
     */
    public BluetoothDevice originalDevice;
    /**
     * 广播数据
     */
    public byte[] scanRecord;
    /**
     * 设备名称
     */
    public String name = "";
    /**
     * 设备id
     */
    public String devId = "";
    /**
     * 设备地址
     */
    public String addr = "";
    /**
     * 固件版本
     */
    public String firmware = "";
    /**
     * 硬件版本
     */
    public String hardware = "";
    /**
     * 设备类型
     */
    public int type = -1;
    /**
     * 电量
     */
    public int battery = -1;
    /**
     * 信号强度
     */
    public int rssi = -1000;
    /**
     * 工作模式
     */
    public int mode;
    /**
     * 连接状态
     */
    public int connectionState = Connection.STATE_DISCONNECTED;
    /**
     * 配对状态
     */
    public int bondState;

    public Device() {
    }

    @Override
    public Device clone() {
        Device device = null;
        try {
            device = (Device) super.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return device;
    }

    public static Device valueOf(@NonNull BluetoothDevice bluetoothDevice) {
        Device device = new Device();
        device.name = bluetoothDevice.getName();
        device.addr = bluetoothDevice.getAddress();
        device.bondState = bluetoothDevice.getBondState();
        device.originalDevice = bluetoothDevice;
        return device;
    }
    
    @Override
    public boolean equals(Object obj) {
        return obj instanceof Device && addr.equals(((Device) obj).addr);
    }

    @Override
    public int hashCode() {
        return addr.hashCode();
    }

    @Override
    public int compareTo(@NonNull Device another) {
        int result;
        if (rssi == 0) {
            return -1;
        } else if (another.rssi == 0) {
            return 1;
        } else {
            result = Integer.compare(another.rssi, rssi);
            if (result == 0) {
                result = name.compareTo(another.name);
            }
        }
        return result;
    }

    public boolean isConnected() {
        return connectionState == Connection.STATE_SERVICE_DISCOVERED;
    }

    public boolean isDisconnected() {
        return connectionState == Connection.STATE_DISCONNECTED || connectionState == Connection.STATE_RELEASED;
    }

    public boolean isConnecting() {
        return connectionState != Connection.STATE_DISCONNECTED && connectionState != Connection.STATE_SERVICE_DISCOVERED &&
                connectionState != Connection.STATE_RELEASED;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(this.originalDevice, flags);
        dest.writeByteArray(this.scanRecord);
        dest.writeString(this.name);
        dest.writeString(this.devId);
        dest.writeString(this.addr);
        dest.writeString(this.firmware);
        dest.writeString(this.hardware);
        dest.writeInt(this.type);
        dest.writeInt(this.battery);
        dest.writeInt(this.rssi);
        dest.writeInt(this.mode);
        dest.writeInt(this.connectionState);
        dest.writeInt(this.bondState);
    }

    protected Device(Parcel in) {
        this.originalDevice = in.readParcelable(BluetoothDevice.class.getClassLoader());
        this.scanRecord = in.createByteArray();
        this.name = in.readString();
        this.devId = in.readString();
        this.addr = in.readString();
        this.firmware = in.readString();
        this.hardware = in.readString();
        this.type = in.readInt();
        this.battery = in.readInt();
        this.rssi = in.readInt();
        this.mode = in.readInt();
        this.connectionState = in.readInt();
        this.bondState = in.readInt();
    }

    public static final Parcelable.Creator<Device> CREATOR = new Parcelable.Creator<Device>() {
        @Override
        public Device createFromParcel(Parcel source) {
            return new Device(source);
        }

        @Override
        public Device[] newArray(int size) {
            return new Device[size];
        }
    };
}

File: library/src/main/java/com/snail/easyble/core/Device.kt
package com.snail.easyble.core

import android.bluetooth.BluetoothDevice
import android.os.Parcel
import android.os.Parcelable

/**
 * Represents a remote Bluetooth device.
 * 
 * date: 2018/4/11 15:00
 * author: zengfansheng
 */
open class Device constructor(val originalDevice: BluetoothDevice) : Comparable<Device>, Cloneable, Parcelable {
    /** Raw bytes of scan record */
    var advData: ByteArray? = null
    var name = ""
    /** Uniquely identifies */
    var devId = ""
    var addr = originalDevice.address!!
        private set
    var firmware = ""
    var hardware = ""
    var battery = -1
    var rssi = -100
    /** One of [IConnection.STATE_DISCONNECTED], [IConnection.STATE_CONNECTING],
     * [IConnection.STATE_SCANNING], [IConnection.STATE_CONNECTED], [IConnection.STATE_SERVICE_DISCOVERING],
     * [IConnection.STATE_SERVICE_DISCOVERED]
     */
    var connectionState = IConnection.STATE_DISCONNECTED
    var bondState = BluetoothDevice.BOND_NONE
    var isConnectable: Boolean? = null
        internal set

    init {
        name = originalDevice.name ?: ""
        bondState = originalDevice.bondState
    }

    val isConnected: Boolean
        get() = connectionState == IConnection.STATE_SERVICE_DISCOVERED

    val isDisconnected: Boolean
        get() = connectionState == IConnection.STATE_DISCONNECTED || connectionState == IConnection.STATE_RELEASED

    val isConnecting: Boolean
        get() = connectionState != IConnection.STATE_DISCONNECTED && connectionState != IConnection.STATE_SERVICE_DISCOVERED &&
                connectionState != IConnection.STATE_RELEASED

    override fun equals(other: Any?): Boolean {
        return other is Device && addr == other.addr
    }

    override fun hashCode(): Int {
        return addr.hashCode()
    }

    override fun compareTo(other: Device): Int {
        var result: Int
        if (rssi == 0) {
            return -1
        } else if (other.rssi == 0) {
            return 1
        } else {
            result = Integer.compare(other.rssi, rssi)
            if (result == 0) {
                result = name.compareTo(other.name)
            }
        }
        return result
    }

    constructor(source: Parcel) : this(source.readParcelable<BluetoothDevice>(BluetoothDevice::class.java.classLoader)!!) {
        readFromParcel(source)
    }
    
    protected fun readFromParcel(source: Parcel) {
        advData = source.createByteArray()
        name = source.readString()!!
        devId = source.readString()!!
        addr = source.readString()!!
        firmware = source.readString()!!
        hardware = source.readString()!!
        battery = source.readInt()
        rssi = source.readInt()
        connectionState = source.readInt()
        bondState = source.readInt()
        isConnectable = when (source.readInt()) {
            1 -> true
            2 -> false
            else -> null
        }
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        dest.writeParcelable(originalDevice, flags)
        dest.writeByteArray(advData)
        dest.writeString(name)
        dest.writeString(devId)
        dest.writeString(addr)
        dest.writeString(firmware)
        dest.writeString(hardware)
        dest.writeInt(battery)
        dest.writeInt(rssi)
        dest.writeInt(connectionState)
        dest.writeInt(bondState)
        dest.writeInt(if (isConnectable == null) 0 else if (isConnectable!!) 1 else 2)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Device> = object : Parcelable.Creator<Device> {
            override fun createFromParcel(source: Parcel): Device = Device(source)
            override fun newArray(size: Int): Array<Device?> = arrayOfNulls(size)
        }
    }
}
